src/db.c

```diff
@@ -121,13 +125,13 @@ robj *lookupKey(redisDb *db, robj *key, int flags) {
             }
         }
 
-        if (!(flags & (LOOKUP_NOSTATS | LOOKUP_WRITE)))
+        if (!(flags & LOOKUP_NOSTATS) && (!(flags & LOOKUP_WRITE) || flags & LOOKUP_WRITE_STATS))
             server.stat_keyspace_hits++;
         /* TODO: Use separate hits stats for WRITE */
     } else {
-        if (!(flags & (LOOKUP_NONOTIFY | LOOKUP_WRITE)))
+        if (!(flags & LOOKUP_NONOTIFY) && (!(flags & LOOKUP_WRITE) || flags & LOOKUP_WRITE_NOTIFY))
             notifyKeyspaceEvent(NOTIFY_KEY_MISS, "keymiss", key, db->id);
-        if (!(flags & (LOOKUP_NOSTATS | LOOKUP_WRITE)))
+        if (!(flags & LOOKUP_NOSTATS) && (!(flags & LOOKUP_WRITE) || flags & LOOKUP_WRITE_STATS))
             server.stat_keyspace_misses++;
         /* TODO: Use separate misses stats and notify event for WRITE */
     }
```

Comment:
On a higher lever, what we're saying here is that it's a write, but it is also a read command, so we want the effect of both a read lookup and a write lookup.

If we ever fix the todos here, i.e. use separate hits/misses counters for writes, it is going to be even more complicated.

So, I have another suggestion: We can introduce a flag `LOOKUP_READWRITE`, which implies `WRITE`.

     *  LOOKUP_READWRITE: Implies LOOKUP_WRITE, but also affects stats and keyspace
     *                    notifications for reads.

(In lookupKey() we can add `if (flags & LOOKUP_READWRITE) flags |= LOOKUP_WRITE;`)

Alternatively, we can introduce LOOKUP_READ, which is implied if WRITE is not given.

     *  LOOKUP_READ: Optional and implied if LOOKUP_WRITE is not given. However, to
     *               get the effect of both read and write, you can specify
     *               LOOKUP_READ | LOOKUP_WRITE

Reply:
since the majority of the code calls lookupKeyRead (and doesn't directly calls lookupKey), maybe we can add an explicit LOOKUP_READ.

Reply:
i.e. i support the idea of `READ|WRITE`.
but i'm not yet certain if this shouldn't be a simple WRITE.
it depends if GETSET is both GET and SET, or same as INCR (returns the counter, but it's only a WRITE command).

i suppose GETSET (works on strings), is used for caching cases, and that the `keymiss` KSN is also meant for that, so maybe getting both read effects and write effects is the desired thing.
@itamarhaber please share your thoughts.

Reply:
The problem with an explicit LOOKUP_READ is what we get more combinations. What to do if neither READ nor WRITE is given? IMO we imply READ in that case.

Reply:
if we want, we can assert that at least one of these flags i set.
but considering `lookupKey` is internal to db.c (and can be made static), maybe that's not necessary.

Reply:
Use an explicit LOOKUP_READ and add this code `if (~flags & LOOKUP_WRITE) flags |= LOOKUP_READ;` to indicate read when there is no LOOKUP_WRITE flag. Is it safe to do this?

Reply:
@jerrykcode Yes I think it's safe. Logically it is the same as what we got with your previous flags (LOOKUP_WRITE_STATS, etc.) isn't it?

I think Oran meant we can set LOOKUP_READ in lookupKeyReadWithFlags() since all key lookups go via the lookupKeyRead or lookupKeyWrite functions, but logically this is the same too. It's a matter preference.

Reply:
@zuiderkwast Yes logically they are the same things. But if we do not set LOOKUP_READ in lookupKey(), then the effect of `calling lookupKey() without any flags` will be different from before. Will this be a problem? Of course we never call lookupKey() directly so this difference doesn't matter in logically.

Reply:
I think it's no problem to put it in lookupKeyReadWithFlags. There is no prototype for lookupKey() so it is only available inside db.c.